### ISSUE TITLE
[Tags] Use simpler "categories/tags" terminology

### ISSUE THREAD

**Issue by tobscure** [flarum/core#395]
Primary tags → categories
Secondary tags → tags

Consider renaming the whole extension to Taxonomy.

**Comment by darkspotinthecorner** [comment]
Maybe changing the flarum/tags extension would work best.
Allow the admins to create any number of taxonomies, each with specific attributes (nested vs. flat, single- vs. multi-assigned) and give them a name. That way the could add flavor to their board by labeling the taxonomies according to their theme and you don't have to deal with the headache of limiting flarum to a specific philosophy. :wink:

**Comment by franzliedke** [comment]
That would be very powerful, but probably also very complicated and hard to understand. Sometimes, too much flexibility can be a pain for the user...

**Comment by jenolan** [comment]
This is a good idea except that it could become confusing (as said above by @franzliedke) 
The current method of a two level primary tag, but allowing the admin to name it as they want (ie Category) and then secondary tags could be names 'Tags' ... what should be available is to have the secondary tags appear below the primary ones to allow this differentiation, see screen shot, if the grey ones were 'indented' from the primary ones voila category / tag handling ;-)

<img width="878" alt="screen shot 2015-09-04 at 12 26 28 pm" src="https://cloud.githubusercontent.com/assets/1505228/9675407/90f9b17a-5300-11e5-98d2-2a8a2edd7ded.png">

**Comment by tobscure** [comment]
> That would be very powerful, but probably also very complicated and hard to understand. Sometimes, too much flexibility can be a pain for the user...

The challenge is to implement a system with that kind of power (for users who want to take advantage of it), but with a simple UI so users who just want plain old categories won't get confused. I think I'm up for the challenge ;)

**Comment by darkspotinthecorner** [comment]
> The challenge is to implement a system with that kind of power (for users who want to take advantage of it), but with a simple UI so users who just want plain old categories won't get confused. I think I'm up for the challenge ;)

Way to go! :)

Maybe a default-UI and the option to switch to an extended-UI.

**Comment by darkspotinthecorner** [comment]
Just thought about this. Mainly how to make it easy to understand and still flexible.

I would think this may be good place for a natural language form, something like [this](http://tympanus.net/Tutorials/NaturalLanguageForm/).

I made a crude writeup of phrases that could be used. Still needs polish, mind you. ;)

```
Part 1 (define taxonomy name):
	A term in this new taxonomy will be called ">>label<<", multiple terms will be called ">>labels<<".

Part 2 (define maximum):
	Items may only belong to a single {"label"}.
	- or -
	Items may only belong to a maximum of ">>5<<" {"labels"}.
	- or -
	Items may belong to any number of {"labels"}.

Part 3 (define minimum):
	New items do not need a {"label"}.
	- or -
	New items need to have at least one {"label"}.
	- or -
	New items need to have at least ">>2<<" {"labels"}.

Part 4 (nesting?):
	The {"labels"} cannot be nested.
	- or -
	The {"labels"} may be nested as many times as needed.
	- or -
	The {"labels"} may be nested, up to a maximum depth of ">>3<<" levels.

Part 5 (viewing items):
	When viewing the items inside a {"label"}, only show the directly assigned items.
	- or -
	When viewing the items inside a {"label"}, show all items in sublevels, too.
```

**Comment by SiCoUK** [comment]
Having just installed the beta to test I have to agree that the primary / secondary tag naming is confusing and I would vote for Categories & Tags (or customisable naming). I would of thought that allowing all sorts of customisation may be over kill!

I would also vote to hide secondary tags on the discussion view unless you are in the primary tag. I can see them becoming extremely crowded.

**Comment by jordanjay29** [comment]
If I can raise this issue from the dead, I'm not sure where this stands, but I honestly think the term 'categories' is tried and tired in the forumsphere. I like tags, labels, groups (as ambiguous as that is), channels, etc.

**Comment by dav-is** [comment]
There MUST be an option for a regularly styled forum layout. I feel the current way Flarum operates is "chatty". I like it but there needs to be an option for bigger comunities that have many subsections.

It would be cool if you could apply the tags in the subsection without that subsection being the primary tag. There could be global tags and subsection specific tags.

For example:

Main forum =>
    Johns Specific Subsection =>
      Post 1 (FAQ->Tutorial)
      Post 2 (Support->Php)
      Post 3 (General)

**Comment by jordanjay29** [comment]
> There MUST be an option for a regularly styled forum layout.

I feel like the option lies in an extension (which, I'll note, that's all that Tags is).

**Comment by Luceos** [comment]
I wouldn't be too surprised to see a categories extension come into life any time in the future. If so, it can perfectly exist next to the core tags extension.

Related to terminology, out of experience I've used so far:

- primary tags
- sub tags, which are child tags of primary tags
- secondary tags or classifiers

Related to renaming the extension. I wouldn't do that. But we might redefine the focus of what the goals of this package should be. Nothing in this package is forced. If we'd leave out the tag-index page we would have a perfectly valid and simple tagging system.

**Comment by dcsjapan** [comment]
I'm starting to really like Toby's initial suggestion.

For those who haven't seen it yet, you can see a graphic representation of what he's got in mind down at the end of his second reply to https://github.com/flarum/core/issues/746.

**Comment by dcsjapan** [comment]
It just occurred to me that we might want to continue using the term "tags" as the generic name for the things handled by this extension. The reason is this:

![tagchange](https://cloud.githubusercontent.com/assets/13129997/14306545/0959f854-fc03-11e5-99e7-a7bcaa6016ef.png)

Rewriting this message to use the "categories/tags" terminology correctly could make things too complex. Another approach might be to keep the "tags" term as the generic, and use another word ... such as "keywords", maybe ... to describe what we're currently calling "secondaries". That gives us:

- Tags
 - Category Tags (a.k.a. "Categories" and "Subcategories")
 - Keyword Tags (a.k.a. "Keywords")

That would allow you to leave the post stream message as it is ... and it would also mean you could get away without renaming the extension to Taxonomy.

**Comment by dcsjapan** [comment]
As for [right-aligning categories in the topic list](https://discuss.flarum.org/d/1144-should-the-categories-in-the-topic-list-be-right-aligned) ... while I agree with that suggestion, I don't much like how the tags fade out when they overflow the space allotted for them:

![tags](https://cloud.githubusercontent.com/assets/13129997/14483686/78363734-0184-11e6-9045-f8232bc998ae.png)

It looks like a UI malfunction rather than a feature. How about showing only the two most important tags instead, and displaying the rest when the user hovers over them? That would require some means of prioritizing them, such as:

- Category tags
- Subcategory tags
- Secondary tags

And if there's a tie for second place (e.g. a thread has one category tag and two secondary tags) it could show whichever secondary tag was chosen first. This wouldn't be very discoverable, but it would be a bit prettier and help save space for the date in the mobile view

**Comment by franzliedke** [comment]
@dcsjapan I agree that making them appear on hover makes sense, but I'm okay with the way it is right now. Did anybody complain on the forums so far?

**Comment by franzliedke** [comment]
Oh wait, they already shift to the left on hover. Nice. :)

**Comment by dcsjapan** [comment]
> I agree that making them appear on hover makes sense, but I'm okay with the way it is right now. Did anybody complain on the forums so far?

No, just me, probably. :package: 

The appearing on hover is good, but the half-faded appearance feels wrong somehow.

**Comment by dav-is** [comment]
I agree it looks weird. Perhaps it shows 2 tags then a '...' At the end

**Comment by dcsjapan** [comment]
While reworking the Tags extension, would it be possible to add some way to fine-tune which tags (or which types of tags) are displayed in the sidebar list? See [this forum post](https://discuss.flarum.org/d/3355-hidden-tags-in-main-list/13) for an example of the sort of control that admins might appreciate having.

**Comment by tobscure** [comment]
@dcsjapan Yes I think that kind of power is probably a good idea.

**Comment by sirinath** [comment]
> Allow the admins to create any number of taxonomies, each with specific attributes (nested vs. flat, single- vs. multi-assigned) and give them a name. That way they could add flavor to their board by labeling the taxonomies according to their theme and you don't have to deal with the headache of limiting flarum to a specific philosophy.

I think you should do this with pre defined defaults so advance uses can have more flexibility.

**Comment by Dremor** [comment]
I'll add two idea : 
- limit some tags to some groups (admins, etc...). Like that you can use tags to assign a status to a new feature (To Do, Completed, Accepting merge request, etc.), which would only be assigned by a project developer.
- Give more level (more than the current 2 levels) to the sidebar. In my case, I would like to have a "Main -> Idea -> Accepted" category.

**Comment by kwiesmueller** [comment]
On the forum it was suggested to post my message here, so there you go:

I just encountered some questions regarding Tags that should be features for me.
First thing:
Allowing certain Roles/Groups to add Tags (create new ones). I can not find any permission allowing that and it would be really annoying to make everybody admin that should be allowed to add tags.
(I did not find any way to make a group admin or at least some kind of admin as well to let them access some settings).

Aside from this it would look very helpful to me, to allow (some) users to create secondary tasks automatically when creating their new post. This would be some pretty common behavior and would in the end result in better searchability (at least if I got the secondary tags idea right).

**Comment by askvortsov1** [comment]
I feel like this has devolved into a conversation about fundamentally redesigning the tags extension. I don't think this is a bad thing, but I would like to note that the original issue of confusing naming still stands.

I propose that we implement Toby's proposed rename, and then, if relevant, reconsider this conversation (which I think would benefit by being split up over several issues). The conversation can continue regardless of what primary/secondary tags (or categories/tags) are called, but the rename will make end user's lives less complicated, and is trivial to execute.

**Comment by luceos** [comment]
I disagree with the original terminology proposed. Categories is absolutely not something that makes this simpler. I would suggest to stick to using the Tag name, but prefix them to make their use and interpretation more straightforward, eg: Primary tags, Sub tags and Helper tags.

Discussion should focus on the terminology here not on changing the actual implementation of the tags extension.

**Comment by askvortsov1** [comment]
At least from my understanding, there are 2 issues at hand here.

1. Is potentially rewriting the tags extension to allow for custom taxonomies. Personally, this is something I don't find necessary: I think it adds too much unnecessary complexity to a system that's already complex enough. Id say that allowing more levels should be considered, although even then I'm not sure how I feel about more than 3 levels at max (too many kinda goes against Flarum's philosophy of simplicity). 

2. What each type of tag should be called. The reason I like categories and tags is that's the naming conventionally used to describe the functionality that primary and secondary tags fulfill. People are already familiar with what a category and tag is, and this isn't just for forum software (blogs  operate similarly)
